elastic/logstash
# Implements DLQ storage policy

## Release notes

Introduce a configuration option to explicitly enable `drop_older` behavior in dead letter queue. 

## What does this PR do?


Exposes `dead_letter_queue.storage_policy` configuration setting to explicitly enable the `drop_older` behavior in DLQs.

Moving from a `drop_newer` to a `drop_older` behavior has impact both on the writer side and to the reader side.
The implementation leverage the fact that a complete DLQ segment can be removed to free up space; on the writer side when the `dead_letter_queue.max_bytes` limit is reached it has to remove old segments. This operation could potentially involve more then one segment, in case the `dead_letter_queue.max_bytes` is downsized, for example between 100 mb to 50 mb.
On the reader side, the consuming has to be adapted to don't expect a continuous flow of segments, it could face an hole due to removal of tail segments.


## Why is it important/What is the impact to the user?


Enabling a `drop_older` behavior has some benefits:
- permit the user to express precedence on fresh events instead of old ones.
- avoid the leak of newer events, and doesn't require to manually remove old segments to let the queue able to process the newer.

## Checklist




## Author's Checklist


- [x] run a local test with a consumer and producer

## How to test this PR locally


- In `config/logstash,yml` set  the policy to `fifo`:
```
dead_letter_queue.storage_policy: fifo
```
- run logstash with a DLQ upstream and downstream, the upstream try to send events to a closed ES index
  -  to close the index 
  ```
  POST test_index/_close
  ```
  - run logstash with:
  ```
  bin/logstash --path.settings "/tmp/dlq_test_config/"
  ```
- it's expected that disabling the downstream pipeline, the DLQ reaches the `dead_letter_queue.max_bytes` no errors log in the form of `cannot write event to DLQ(path` appears, and once re-started the downstream pipeline, there is  a miss of older events.

In dlq_test_config add:
- the `logstash.yml` with the flag
- the `pipelines.yml` pointing to the DLQ upstream and downstream
- the pipelines config

```
- pipeline.id: dlq_upstream
  path.config: "/tmp/dlq_test_config/dlq_upstream_pipeline.conf"

- pipeline.id: dlq_downstream
  path.config: "/tmp/dlq_test_config/dlq_reader_pipeline.conf"
```

```
input {
  generator {
    message => '{"name": "Andrea"}'
    codec => json
  }
}

output {
  elasticsearch {
    index => "test_index"
    hosts => "http://localhost:9200"
    user => "elastic"
    password => "changeme"
  }
}
```

```
input {
  dead_letter_queue {
  	path => "/<LS HOME>/data/dead_letter_queue/"
  	pipeline_id => "dlq_upstream"
  }
}

output {
  stdout {
    codec => dots
  }
}
```


## Related issues


- Relates #13915

## Use cases
